README.md
<div align="center">

<img src="public/assets/images/teampass-logo2-login.png" width="110" alt="Teampass" />

# Teampass

### Self-hosted password management your whole team can trust

Folder-level access control · authenticated AES-256-GCM encryption · compliance evidence<br />
**Your secrets never leave your infrastructure.**

**🌐 [teampass.net](https://teampass.net)** ·
**📖 [Documentation](https://documentation.teampass.net)** ·
**💬 [Discussions](https://github.com/nilsteampassnet/TeamPass/discussions)** ·
**🐳 [Docker Hub](https://hub.docker.com/r/teampass/teampass)**

<br />

[![Release](https://img.shields.io/github/v/release/nilsteampassnet/TeamPass?style=for-the-badge&color=24c8ff&labelColor=0f2740)](https://github.com/nilsteampassnet/TeamPass/releases/latest)
[![License](https://img.shields.io/github/license/nilsteampassnet/TeamPass?style=for-the-badge&color=24c8ff&labelColor=0f2740)](LICENSE.md)
[![PHP](https://img.shields.io/badge/PHP-8.2%2B-24c8ff?style=for-the-badge&labelColor=0f2740&logo=php&logoColor=white)](https://www.php.net/)
[![Docker Pulls](https://img.shields.io/docker/pulls/teampass/teampass?style=for-the-badge&color=24c8ff&labelColor=0f2740&logo=docker&logoColor=white)](https://hub.docker.com/r/teampass/teampass)
[![Stars](https://img.shields.io/github/stars/nilsteampassnet/TeamPass?style=for-the-badge&color=24c8ff&labelColor=0f2740)](https://github.com/nilsteampassnet/TeamPass)

[![CodeQL](https://img.shields.io/github/actions/workflow/status/nilsteampassnet/TeamPass/codeql.yml?branch=master&style=for-the-badge&label=CodeQL&labelColor=0f2740&color=3fb950&logo=github&logoColor=white)](https://github.com/nilsteampassnet/TeamPass/actions/workflows/codeql.yml)
[![Docker Build](https://img.shields.io/github/actions/workflow/status/nilsteampassnet/TeamPass/docker-publish.yml?branch=master&style=for-the-badge&label=Docker%20Build&labelColor=0f2740&color=3fb950&logo=docker&logoColor=white)](https://github.com/nilsteampassnet/TeamPass/actions/workflows/docker-publish.yml)
[![Security policy](https://img.shields.io/badge/Security-policy_%26_advisories-3fb950?style=for-the-badge&labelColor=0f2740&logo=shieldsdotio&logoColor=white)](https://github.com/nilsteampassnet/TeamPass/security/advisories)
[![Sponsor](https://img.shields.io/github/sponsors/nilsteampassnet?style=for-the-badge&label=Sponsors&labelColor=0f2740&color=ff4dda&logo=githubsponsors&logoColor=ff4dda)](https://github.com/sponsors/nilsteampassnet)

</div>

---

## Contents

- [About](#about)
- [Who it's for](#who-its-for)
- [Security](#security)
- [Features](#features)
- [Get started](#get-started)
- [Documentation](#documentation)
- [Languages](#languages)
- [Community](#community)
- [Support Teampass](#support-teampass)
- [License](#license)

---

## About

**Teampass is an open-source credential vault you run yourself.** No account to create, no company behind the curtain holding your data — just a PHP/MySQL application on your own server, with folder-level access control, per-user encryption keys and a full audit trail.

It has been built and maintained since 2009, driven by what real teams actually run into: who should see which credential, how to prove it to an auditor, and how to stop passwords living in chat threads and spreadsheets.

<div align="center">
  <img src="https://teampass.net/images/portfolio/tp3_sw_1.png" width="820" alt="Teampass interface" />
</div>

<details>
<summary><b>📸 More screenshots</b></summary>
<br />

**Items and secrets**

<img src="docs/_media/tp3_items_1.png" width="420" alt="Item list" />
<img src="docs/_media/tp3_items_2.png" width="420" alt="Item detail" />

**Folders and roles**

<img src="docs/_media/tp3_subfolders_1.png" width="420" alt="Folder tree" />
<img src="docs/_media/tp3_features_roles_1.png" width="420" alt="Roles" />
<img src="docs/_media/tp3_features_roles_3.png" width="420" alt="Role rights" />
<img src="docs/_media/tp3_features_roles_5.png" width="420" alt="Role assignment" />

**Authentication and MFA**

<img src="docs/_media/tp3_auth_mfa_1.png" width="420" alt="MFA setup" />
<img src="docs/_media/tp3_auth_oauth2_1.png" width="420" alt="OAuth2 settings" />

**Encryption keys**

<img src="docs/_media/tp3_keys_1.png" width="420" alt="Key management" />
<img src="docs/_media/tp3_keys_5.png" width="420" alt="Key regeneration" />

**Search, export, one-time view**

<img src="docs/_media/tp3_settings_keyword_search.png" width="420" alt="Keyword search" />
<img src="docs/_media/tp3_export_1.png" width="420" alt="Export" />
<img src="docs/_media/tp3_otv_1.png" width="420" alt="One-time view" />
<img src="docs/_media/tp3_otp_1.png" width="420" alt="TOTP" />

**Background tasks**

<img src="docs/_media/tp3_tasks_04.png" width="420" alt="Tasks" />
<img src="docs/_media/settings_tasks_options_01.png" width="420" alt="Task settings" />

</details>

---

## Who it's for

<table>
<tr>
<td width="33%" valign="top">

### 🏠 Individuals & Homelab

***Own your vault, literally.***

- Runs on a Raspberry Pi or a €5 VPS
- Personal folders encrypted with your own key
- Import from Bitwarden, LastPass, 1Password or KeePassXC
- Free forever, no sign-up required

</td>
<td width="33%" valign="top">

### 👥 Teams & SMB

***Stop sharing passwords in chat.***

- Folders and roles instead of a shared document
- A record of who accessed what, and when
- Secure Send for clients and contractors
- Browser extension for day-to-day autofill

</td>
<td width="33%" valign="top">

### 🏛️ Enterprise & Regulated

***Prove your access controls, don't just claim them.***

- Access recertification campaigns with immutable decisions
- Compliance reports and evidence export
- LDAP/AD with nested groups, OAuth2 SSO
- Data classification and ownership

</td>
</tr>
</table>

---

## Security

### Encryption you can describe to an auditor

Secrets are encrypted with **AES-256-GCM** using random nonces and per-secret salts, under **256-bit object keys**. The private key that unlocks them is derived from your password with **PBKDF2-SHA256 at 600 000 iterations**.

- **Authenticated encryption** — tampering is detected, not silently decrypted - **Per-user key distribution** — every user holds their own RSA-wrapped copy of each object key, so removing an account actually revokes access instead of just hiding a button
- **Lazy migration** — format upgrades happen on access, with no maintenance window

<div align="center">
  <img src="docs/_media/tp3_encryption_model.webp" width="700" alt="Teampass encryption model" />
</div>

### Transparency over silence

> A password manager that reports no vulnerabilities is not a password manager that has none.

Findings are triaged, fixed and published as GitHub Security Advisories with CVE identifiers.

- 🔒 [Security policy and how to report](SECURITY.md)
- 📋 [Published advisories](https://github.com/nilsteampassnet/TeamPass/security/advisories)
- 🛡️ [Security hardening guide](https://documentation.teampass.net/#/install/security-hardening)

**Found a vulnerability?** Please report it privately through [GitHub Security Advisories](https://github.com/nilsteampassnet/TeamPass/security/advisories/new) — never in a public issue.

---

## Features

<details>
<summary><b>🗂️ Folder and role access control</b></summary>
<br />

- [Folders](https://documentation.teampass.net/#/features/folders) — unlimited nesting, per-folder password complexity rules
- [Roles](https://documentation.teampass.net/#/features/roles) — grant access by role, not user by user
- [Rights](https://documentation.teampass.net/#/features/rights) — read, write, no-edit, no-delete, resolved least-permissive-wins
- [Users](https://documentation.teampass.net/#/features/users) — per-user overrides on top of roles
- Personal folders that nobody else can read, including administrators

</details>

<details>
<summary><b>🔐 Authenticated encryption</b></summary>
<br />

- [Encryption model](https://documentation.teampass.net/#/install/encryption) — how keys are derived, wrapped and distributed
- [Encryption improvements](https://documentation.teampass.net/#/install/encryption-improvements) — the AES-256-GCM format and its lazy migration
- [Key management](https://documentation.teampass.net/#/features/keys) — regeneration, recovery, per-user key repair

</details>

<details>
<summary><b>📊 Security posture</b></summary>
<br />

- Security Posture Dashboard scoring weak, reused and breached credentials
- [Breach detection](https://documentation.teampass.net/#/features/breach-detection) — Have I Been Pwned checks without sending your passwords
- [Password renewal](https://documentation.teampass.net/#/features/renewal) — expiry policies and reminders
- [Micro-learning](https://documentation.teampass.net/#/features/micro-learning) — in-app nudges instead of a yearly slideshow

</details>

<details>
<summary><b>🏛️ Governance and audit</b></summary>
<br />

- [Access reviews](https://documentation.teampass.net/#/manage/access-reviews) — recertification campaigns with immutable decisions
- [Compliance reports](https://documentation.teampass.net/#/manage/compliance-reports) — evidence you can hand to an auditor
- [Rotation tracking](https://documentation.teampass.net/#/manage/rotation-tracking) — what was rotated, when, by whom
- [Leaver risk](https://documentation.teampass.net/#/features/leaver-risk) — what a departing user could still know
- [Data classification](https://documentation.teampass.net/#/features/classification) — ownership and sensitivity labels

</details>

<details>
<summary><b>🪪 Identity integration</b></summary>
<br />

- [Authentication](https://documentation.teampass.net/#/features/authentication) — local, LDAP/AD with nested groups, OAuth2 / SSO
- Multi-factor: TOTP (Google Authenticator), Duo Security, YubiKey, AGSES
- [Network ACL](https://documentation.teampass.net/#/manage/network-acl) — restrict access by IP range
- [Session management](https://documentation.teampass.net/#/misc/session-management) — timeouts, concurrent sessions, Redis-backed storage

</details>

<details>
<summary><b>🤖 Automation</b></summary>
<br />

- [REST API](https://documentation.teampass.net/#/api/api-basic) — JWT-authenticated, OpenAPI 3.1 spec, Bash and PowerShell clients
- [Browser extension](https://documentation.teampass.net/#/misc/extension) — autofill, capture, one-click auto-configuration
- [Background tasks](https://documentation.teampass.net/#/manage/tasks) — key distribution, notifications, maintenance
- [Real-time collaboration](https://documentation.teampass.net/#/features/collaboration) — WebSocket sync and edition locks
- [Backups](https://documentation.teampass.net/#/features/backups) — scheduled, encrypted database dumps

</details>

<details>
<summary><b>📦 Migration in and out</b></summary>
<br />

- [Import](https://documentation.teampass.net/#/features/import) — Bitwarden, LastPass, 1Password, KeePassXC, CSV
- [Export](https://documentation.teampass.net/#/features/export) — CSV, PDF, and a self-contained encrypted offline HTML vault
- No lock-in: it is your database, on your server, under GPL-3.0

</details>

<details>
<summary><b>⚡ Daily productivity</b></summary>
<br />

- [Search](https://documentation.teampass.net/#/features/search) — across labels, descriptions, tags and custom fields
- [Command palette](https://documentation.teampass.net/#/features/command-palette) — keyboard-first navigation
- [Favourites](https://documentation.teampass.net/#/features/favourites) and quick access to recent items
- [Custom fields](https://documentation.teampass.net/#/features/custom-fields) — individually encryptable, role-restricted
- [Knowledge base](https://documentation.teampass.net/#/features/knowledge-base) and [notification center](https://documentation.teampass.net/#/features/notification-center)
- One-time view links and Secure Send for sharing outside the vault

</details>

---

## Get started

### Requirements

| | |
|---|---|
| **Database** | MySQL 5.7+ or MariaDB 10.7+ |
| **PHP** | 8.2 or newer (tested against 8.3) |
| **Required extensions** | `openssl` `mysqli` `mbstring` `bcmath` `iconv` `xml` `gd` `curl` `gmp` — plus `ldap` for LDAP/AD |
| **Recommended extensions** | `apcu` (config cache) · `opcache` (performance) · `redis` (HA sessions) · `pcntl` + `posix` (WebSocket daemon) |

Teampass follows active PHP support. Running the latest stable PHP release is strongly recommended for both security and performance.

### 🐳 Docker

```bash
docker run -d --name teampass \
  -p 8080:80 \
  -v teampass_data:/var/www/html \
  teampass/teampass:latest
```

Images are published to both registries:

- Docker Hub — `teampass/teampass`
- GitHub Container Registry — `ghcr.io/nilsteampassnet/teampass`

📖 [Docker guide](docs/DOCKER.md) · [Migrating an existing install to Docker](docs/DOCKER-MIGRATION.md)

### 🖥️ Bare metal (recommended for production)

Installing directly on a PHP/MySQL server gives the best performance and the most control over your environment.

- 📖 [Official installation guide](https://documentation.teampass.net/#/install/installation)
- 🔄 [Upgrade guide](https://documentation.teampass.net/#/install/upgrade)
- 🔑 [File permissions](https://documentation.teampass.net/#/install/file-permissions)
- 🎥 [Video tutorial](https://youtu.be/eXieWAIsGzc?feature=shared)

---

## Documentation

| | |
|---|---|
| 📖 **[Full documentation](https://documentation.teampass.net)** | Install, features, administration |
| 🚀 **[Installation](https://documentation.teampass.net/#/install/installation)** | Step-by-step first setup |
| 🔄 **[Upgrade](https://documentation.teampass.net/#/install/upgrade)** | Moving between versions |
| 🛡️ **[Security hardening](https://documentation.teampass.net/#/install/security-hardening)** | Production checklist |
| ⚙️ **[Performance](https://documentation.teampass.net/#/install/performance)** | PHP-FPM, caching, tuning |
| 🔌 **[REST API](https://documentation.teampass.net/#/api/api-basic)** | Endpoints, JWT auth, clients |
| 🧩 **[Browser extension](https://documentation.teampass.net/#/misc/extension)** | Setup and usage |
| 🩺 **[Troubleshooting](https://documentation.teampass.net/#/misc/troubleshooting)** | When something goes wrong |

---

## Languages

Teampass ships in **25 languages**, translated by the community.

<table>
<tr>
<td><img src="public/assets/images/flags/us.png" width="16" alt="" /> English</td>
<td><img src="public/assets/images/flags/fr.png" width="16" alt="" /> French</td>
<td><img src="public/assets/images/flags/de.png" width="16" alt="" /> German</td>
<td><img src="public/assets/images/flags/es.png" width="16" alt="" /> Spanish</td>
<td><img src="public/assets/images/flags/it.png" width="16" alt="" /> Italian</td>
</tr>
<tr>
<td><img src="public/assets/images/flags/pr.png" width="16" alt="" /> Portuguese</td>
<td><img src="public/assets/images/flags/br.png" width="16" alt="" /> Portuguese (BR)</td>
<td><img src="public/assets/images/flags/nl.png" width="16" alt="" /> Dutch</td>
<td><img src="public/assets/images/flags/ru.png" width="16" alt="" /> Russian</td>
<td><img src="public/assets/images/flags/ua.png" width="16" alt="" /> Ukrainian</td>
</tr>
<tr>
<td><img src="public/assets/images/flags/pl.png" width="16" alt="" /> Polish</td>
<td><img src="public/assets/images/flags/cz.png" width="16" alt="" /> Czech</td>
<td><img src="public/assets/images/flags/hu.png" width="16" alt="" /> Hungarian</td>
<td><img src="public/assets/images/flags/ro.png" width="16" alt="" /> Romanian</td>
<td><img src="public/assets/images/flags/bg.png" width="16" alt="" /> Bulgarian</td>
</tr>
<tr>
<td><img src="public/assets/images/flags/gr.png" width="16" alt="" /> Greek</td>
<td><img src="public/assets/images/flags/tr.png" width="16" alt="" /> Turkish</td>
<td><img src="public/assets/images/flags/se.png" width="16" alt="" /> Swedish</td>
<td><img src="public/assets/images/flags/no.png" width="16" alt="" /> Norwegian</td>
<td><img src="public/assets/images/flags/ee.png" width="16" alt="" /> Estonian</td>
</tr>
<tr>
<td><img src="public/assets/images/flags/catalonia.png" width="16" alt="" /> Catalan</td>
<td><img src="public/assets/images/flags/cn.png" width="16" alt="" /> Chinese</td>
<td><img src="public/assets/images/flags/ja.png" width="16" alt="" /> Japanese</td>
<td><img src="public/assets/images/flags/vi.png" width="16" alt="" /> Vietnamese</td>
<td>🇸🇦 Arabic <sub>(in progress)</sub></td>
</tr>
</table>

Translations are managed on POEditor — a few strings from you go a long way.

[![Help translate](https://img.shields.io/badge/Help_translate-POEditor-24c8ff?style=for-the-badge&labelColor=0f2740)](https://poeditor.com/join/project?hash=0vptzClQrM)

---

## Community

### Contributing

Contributions of any kind are very welcome. Fork the repo, make your changes, then open a pull request — see [CONTRIBUTING.md](.github/CONTRIBUTING.md) for the development setup, coding standards and branch conventions, and [CODE_OF_CONDUCT.md](CODE_OF_CONDUCT.md) for how we work together.

[![Submit a PR](https://img.shields.io/badge/Submit_a_PR-GitHub-%23060606?style=for-the-badge&logo=github&logoColor=fff)](https://github.com/nilsteampassnet/TeamPass/compare)

### Reporting bugs

If something does not work as it should, raise a ticket. Please include the steps to reproduce, your server configuration, and the diagnostic report from **Admin → Bug icon (bottom left)**.

[![Raise an Issue](https://img.shields.io/badge/Raise_an_Issue-GitHub-%23060606?style=for-the-badge&logo=github&logoColor=fff)](https://github.com/nilsteampassnet/TeamPass/issues/new/choose)

### Asking questions

Questions, deployment advice and ideas belong in Discussions rather than the issue tracker.

[![Discussions](https://img.shields.io/badge/Ask_a_question-Discussions-%23060606?style=for-the-badge&logo=github&logoColor=fff)](https://github.com/nilsteampassnet/TeamPass/discussions)

### Contributors

Teampass is what it is thanks to these people.

[![Contributors](https://readme-contribs.as93.net/contributors/nilsteampassnet/TeamPass?perRow=12&shape=squircle)](https://github.com/nilsteampassnet/TeamPass/graphs/contributors)

### Star history

[![Star History Chart](https://api.star-history.com/svg?repos=nilsteampassnet/TeamPass&type=Date)](https://star-history.com/#nilsteampassnet/TeamPass&Date)

---

## Support Teampass

Teampass is free, GPL-3.0, and has been maintained by one person since 2009. There is no company behind it — which is exactly the point, and also why support matters.

### What sponsorship funds

- **Security work** — triaging reports, fixing them, and publishing advisories with CVEs
- **Releases** — testing, upgrade paths, and keeping older installations able to move forward
- **Documentation** — the guides at documentation.teampass.net
- **Translations** — coordinating 25 languages
- **Infrastructure** — Docker images, CI, and the project websites

### Become a sponsor

The goal is **100 monthly sponsors**. Every tier helps, and small recurring amounts help most because they make the work predictable.

[![Sponsor on GitHub](https://img.shields.io/badge/Sponsor_on_GitHub-nilsteampassnet-%23ff4dda?style=for-the-badge&logo=githubsponsors&logoColor=ff4dda)](https://github.com/sponsors/nilsteampassnet)
[![Donate via PayPal](https://img.shields.io/badge/One--off_donation-PayPal-00457C?style=for-the-badge&logo=paypal&logoColor=fff)](https://www.paypal.com/donate/?hosted_button_id=XUVWYJ7J92X6L)

### Sponsors

Huge thanks to everyone who sponsors this work — you keep Teampass free for everyone else.

[![Sponsors](https://readme-contribs.as93.net/sponsors/nilsteampassnet?perRow=12&shape=squircle)](https://github.com/sponsors/nilsteampassnet)

### Commercial support

Sponsorship funds the free server. The browser extension and professional services fund the roadmap.

| | |
|---|---|
| **Pro Extension** | Browser autofill, capture and phishing protection — from €49/year, 30-day trial |
| **Services** | Feature development, deployment assistance, priority handling — quoted per engagement |

[![Pricing](https://img.shields.io/badge/See_pricing-teampass.net-24c8ff?style=for-the-badge&labelColor=0f2740)](https://teampass.net/pricing.html)

---

## License

Teampass is released under the **[GNU General Public License v3.0](LICENSE.md)**.

You are free to use, study, modify and redistribute it — including commercially — provided derivative works remain under the same license.

📋 [Dependency license compliance report](LICENSE_COMPLIANCE_REPORT.md)

---

<div align="center">

<img src="https://teampass.net/images/madeinfrance.png" width="90" alt="Made in France" />

**Built and maintained by [Nils Laumaillé](mailto:nils@teampass.net) since 2009.**

Copyright © 2009-2026 Nils Laumaillé

<br />

[![StandWithUkraine](https://raw.githubusercontent.com/vshymanskyy/StandWithUkraine/main/badges/StandWithUkraine.svg)](https://github.com/vshymanskyy/StandWithUkraine/blob/main/docs/README.md)

</div>


changelog.md
2.1.27
11/
 Changed licensing to GNU GPL-3.0
 New - User must provide a reason to access a restricted item
 New - Add option to have local and remote accounts when LDAP is enabled
 Improved security of password generator with php7
 Improved cannot edit user without email
 Improved read-only user limitation to copy folder and import action
 Improved tree rebuild with API on folders change
 Improved tables primary and index usage
 Improved LDAP new user by default role
 Improved single click on key icon in items search list
 Improved visibility of path in items list result
 Improved email body with item path
 Introduced an API key by user
 Fix for API keys truncated
 Fix offline password dispay in case of html tags similar in password
 Fix failed folder creation in case of password complexity not reached
 Fix missing quick icons in search results
 #2175 Apostrophes are not handled correctly in usernames
 #2174 Offline mode file bypass read right restrictions
 #2172 2FA Reset Link Can Be Abused
 #2168 API for adding users is not working
 #2167 Info tab is not working if behind a proxy
 #2161 Missing backslash in acount name
 #2160 Added a test for preventing Folders list not shown
 #2154 Personal saltkey is not stored when option enabled
 #2153 [{"error":"no_key_provided"} when running backup script on teampass container
 #2152 No search result and empty popup appear
 #2151 Error in knowledge base that does not show option to swap pages
 #2140 Moving subfolder to root level not possible
 #2127 Grant access with simple folder copy
 #2118 Empty user at Keepass file is not empty after import
 #2116 Insufficient password strength when creating Offline Mode
 #2115 Fix script backup issue with encryption key
 #2111 Add support for login through http header
 #2109 restrict login to Group Ldap don't work
 #2102 Changed field renewal_period size
 #2096 Offline mode decryption fails if too much items exported
 #2095 Can't upload files on items - Plupload update
 #2094 PHP 7.2: Call to undefined function mcrypt_encrypt()
 #2093 role human resources doesn't access expected pages
 #2090 On folder copy, an empty message box is shown
 #2087 Custom fields displayed if empty
 #2085 CSV import error if URL too long
 #2082 API: new folder is allocated to same roles as its parent
 #2081 LDAP bug, can't check connection
 #2080 Email sent on password copy in the clipboard
 #2078 Purge Logs not selection not working properly
 #2077 API: Deleting non existing folder deletes all folders
 #2075 Button "delete selected Items" doesn't work
 #2074 Backup by script not working
 #2073 Move selected Items button
 #2071 Search not working for items with multiple "Restricted to:" roles selected
 #2069 Copy a read-only folder to a read-only folder
 #2066 Read Only Users can Drag and drop items to there personal folder
 #2065 send GAUTH-code on login form doesn't warn user if no email is set
 #2064 LDAP and DUO users with administrator role taken to items page then logged out
 #2063 Setting "Hide forgot password link on Home page" not correctly displayed
 #2059 Fix for Custom LDAP port and adLDAP
 #2035 After first time login as user Personal Folder is not correctly shown (100000)
 #2015 Double click to edit not working for items created by a different user

10/
 Copy password/login button are correctly hidden and shown
 Cursor is hidden on password display to permit a better visualization
 Fixed error while coping an item
 Fixed last 10 items seen list not display on page load
 Fixed display strings with quotes issue
 Fixed page About in Administration page
 Fixed issue when sending email on visualization
 Fixed issue while CSV import in personal sub-folder
 Fixed potential error on item creation without password
 Fixed "show description" option was not taken into consideration
 Fixed an issue in auto-generating csrfp.config file
 Fixed a potential error while generating tp.config file
 Fixed issue in Duo Security log
 Fixed no item details shown on Find page
 Fixed issue related to item edition on doubleclick
 Fixed issue no items shown on tag search
 Fixed issue special characters bad display in search results
 Fixed issue in password clipboard copy with double-quote character
 Added SMTPAutoTLS for sendMail
 Improved offline file with full encryption
 Improved session validity check
 Improved items draggability sometime broken
 Improved personal item edition by enabling the folder edition
 Ensure session extension is always positive
 Added user group allowed for LDAP Windows AD
 Added option to remove the forgot password link on home page if LDAP enabled
 Added option lowercase letters for password generator
 #2058 Importing KeyPass items into private personal folders
 #2056 Custom Fields - displayed in unassigned folders
 #2055 restricted_to of cache table has 2 different values
 #2054 Alter table statement is run only once
 #2053 Custom Fields - folders unchecked at second call up
 #2051 2FA Google Name Displayed field not saving to configuration
 #2043 LDAP Authentication is not working
 #2029 Fix on install step3
 #2028 Fix on install step3
 #2027, #2023 Fatal Error after installation and setting change
 #2025 LDAP allowed user group doesn't work
 #2016 Not all roles visible
 #2013 Bad condition in upgrade script
 #2010 Default Administrated by for user created via LDAP authentication
 #2004 Administrator has no access to Items
 #1997 Error on folder creation of no Parent is selected
 #1996 Some settings resets to defaults
 #1993 Policy for personal saltkey
 #1990 Improving mysql indexes
 #1989 Authentication problem
 #1980 personal item is not deleted
 #1974 Changing saltkey operation
 #1970 ip-api.com use is not compatible with https
 #1960 Access Denied for personal folder
 #1840 add "download" attribute to export file

 9/
 Fixed a possible XSS (credit to ADLab of Venustech)
 Improved security related to User Management
 	> a manager could potentially act on users not related to him

kb.php
<?php
/**
 * @file          kb.php
 * @author        Nils Laumaillé
 * @version       2.1.27
 * @copyright     (c) 2009-2018 Nils Laumaillé
 * @licensing     GNU GPL-3.0
 * @link          http://www.teampass.net
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */

if (!isset($_SESSION['CPM']) || $_SESSION['CPM'] != 1 ||
    !isset($_SESSION['user_id']) || empty($_SESSION['user_id']) ||
    !isset($_SESSION['key']) || empty($_SESSION['key'])
) {
    die('Hacking attempt...');
}

// Load config
if (file_exists('../includes/config/tp.config.php')) {
    require_once '../includes/config/tp.config.php';
} elseif (file_exists('./includes/config/tp.config.php')) {
    require_once './includes/config/tp.config.php';
} else {
    throw new Exception("Error file '/includes/config/tp.config.php' not exists", 1);
}

if (isset($SETTINGS['enable_kb']) === false || $SETTINGS['enable_kb'] !== "1") {
    die('Hacking attempt...');
}

/* do checks */
require_once $SETTINGS['cpassman_dir'].'/sources/checks.php';
if (!checkUser($_SESSION['user_id'], $_SESSION['key'], curPage())) {
    $_SESSION['error']['code'] = ERR_NOT_ALLOWED; //not allowed page
    include $SETTINGS['cpassman_dir'].'/error.php';
    exit();
}

//load language
require_once $SETTINGS['cpassman_dir'].'/sources/main.functions.php';


// Include files
require_once $SETTINGS['cpassman_dir'].'/includes/libraries/protect/SuperGlobal/SuperGlobal.php';
$superGlobal = new protect\SuperGlobal\SuperGlobal();

// Prepare GET variables
$get_id = $superGlobal->get("id", "GET");

//build list of categories
$tab_users = array();
$rows = DB::query(
    "SELECT id, login FROM ".prefix_table("users")." ORDER BY login ASC"
);
$counter = DB::count();
if ($counter > 0) {
    foreach ($rows as $reccord) {
        $tab_users[$reccord['login']] = array(
            'id'=>$reccord['id'],
            'login'=>$reccord['login']
        );
    }
}

echo '
<div class="title ui-widget-content ui-corner-all">
    '.$LANG['knowledge_base'].'&nbsp;&nbsp;&nbsp;
    <button title="'.$LANG['add_new_kb'].'" onclick="OpenDialog(\'kb_form\')" class="button" style="font-size:16px;">
        <i class="fa fa-plus"></i>
    </button>
</div>';

//Show the KB in a table view
echo '
<div style="margin:10px auto 25px auto;min-height:250px;" id="kb_page">
<table id="t_kb" cellspacing="0" cellpadding="5" width="100%">
    <thead><tr>
        <th style="width:50px;"></th>
        <th style="width:25%;">'.$LANG['category'].'</th>
        <th style="width:50%;">'.$LANG['label'].'</th>
        <th style="width:15%;">'.$LANG['author'].'</th>
    </tr></thead>
    <tbody>
        <tr><td></td></tr>
    </tbody>
</table>
</div>';

//Hidden things
echo '
<input type="hidden" id="kb_id" value="" />';

/* DIV FOR ADDING A KN */
echo '
<div id="kb_form" style="display:none;">
    <label for="kb_label" class="label">'.$LANG['label'].'</label>
    <input type="text" id="kb_label" class="input text ui-widget-content ui-corner-all" />
    <br />

    <div style="width:100%;">
        <div style="float:left;width:50%;">
            <label for="kb_category" class="label">'.$LANG['category'].'</label>
            <input name="kb_category" id="kb_category" class="kb_text ui-widget-content ui-corner-all" style="width: 300px;" value="" />
        </div>
        <div style="float:right;width:50%;">
            <label class="modify_kb_label">'.$LANG['anyone_can_modify_it'].' : </label>
            <span class="div_radio">
                <input type="radio" id="modify_kb_yes" name="modify_kb" value="1" checked="checked" /><label for="modify_kb_yes">'.$LANG['yes'].'</label>
                <input type="radio" id="modify_kb_no" name="modify_kb" value="0" /><label for="modify_kb_no">'.$LANG['no'].'</label>
            </span>
        </div>
    </div>

    <div style="float:left;width:100%;">
        <label for="kb_description" class="label">'.$LANG['description'].'</label>
        <textarea rows="5" name="kb_description" id="kb_description" class="input" cols="70"></textarea>
    </div>

    <div style="float:left;width:100%;margin-top:15px;">
        <label for="kb_associated_to" class="label">'.$LANG['associate_kb_to_items'].'</label>
        <select id="kb_associated_to" class="multiselect" multiple="multiple" name="kb_associated_to[]" style="width: 860px; height: 150px;">';
//get list of available items
$items_id_list = array();
if (empty($_SESSION['list_folders_limited'])) {
    $_SESSION['list_folders_limited'] = array();
}
$rows = DB::query(
    "SELECT i.id as id, i.restricted_to as restricted_to, i.perso as perso, i.label as label,
        i.description as description, i.pw as pw, i.login as login, i.anyone_can_modify as anyone_can_modify,
        l.date as date
    FROM ".prefix_table("items")." as i
    INNER JOIN ".prefix_table("log_items")." as l ON (i.id = l.id_item)
    WHERE i.inactif = %i
    AND (l.action = %s OR (l.action = %s AND l.raison LIKE %s))
    AND i.id_tree IN %ls
    ORDER BY i.label ASC, l.date DESC",
    '0',
    'at_creation',
    'at_modification',
    'at_pw :%',
    array_unique(
        array_merge(
            $_SESSION['all_non_personal_folders'],
            $_SESSION['list_folders_editable_by_role'],
            $_SESSION['list_restricted_folders_for_items'],
            $_SESSION['list_folders_limited']
        )
    )
);
foreach ($rows as $reccord) {
    if (!in_array($reccord['id'], $items_id_list) && !empty($reccord['label'])) {
        // exclude item if it is restricted to a group the user doesn't have
        $include_item = false;
        if (empty($reccord['restricted_to'])) {
            $include_item = true;
        } elseif (count(array_intersect(explode(";", $reccord['restricted_to']), $_SESSION['user_roles'])) !== 0) {
            $include_item = true;
        }
        if ($include_item === true) {
            echo '
            <option value="'.$reccord['id'].'">'.$reccord['label'].'</option>';
            array_push($items_id_list, $reccord['id']);
        }
    }
}
echo '
        </select>
    </div>
</div>';

//DELETE DIALOG
echo '
<div id="div_kb_delete" style="display:none;">
    <p><span class="ui-icon ui-icon-alert" style="float:left; margin:0 7px 20px 0;">&nbsp;</span>'.$LANG['confirm_deletion'].'</p>
</div>';

//Hidden things
echo '
<input type="hidden" id="kb_id" value="" />';

//Call javascript stuff
require_once 'kb.load.php';

//If redirection is done to a speoific KB then open it
if ($get_id !== null && empty($get_id) === false) {
    echo '
        <script language="javascript" type="text/javascript">
        <!--
        openKB('.$get_id.');
        -->
        </script>';
}
